src/main/scala/chiseltest/internal/BackendInterface.scala

```
@@ -2,111 +2,93 @@
 
 package chiseltest.internal
 
-import chiseltest.Region
 import chisel3._
 
-import scala.collection.mutable
+import scala.math.BigInt
 
-trait AbstractTesterThread
+/** SimulatorInterface usages in ThreadedBackend. */
+trait BackendInterface { this: ThreadedBackend =>
 
-class TesterThreadList(protected val elts: Seq[AbstractTesterThread]) {
-  def toSeq(): Seq[AbstractTesterThread] = elts
-
-  def join() {
-    Context().backend.doJoin(elts, None)
-  }
-
-  def joinAndStep(clock: Clock) {
-    Context().backend.doJoin(elts, Some(clock))
-  }
-
-  def ++(others: TesterThreadList): TesterThreadList = {
-    new TesterThreadList(elts ++ others.elts)
-  }
-
-  val fork: ForkBuilder = new ForkBuilder(None, None, elts)
-}
-
-/** Common interface definition for tester backends. Internal API.
-  */
-trait BackendInterface {
+  def getClock(clk: Clock): Boolean =
+    simulatorInterface.peek(dataNames(clk)) match {
+      case Some(x) if x == BigInt(1) => true
+      case _                         => false
+    }
 
   /** Writes a value to a clock.
     */
-  def pokeClock(signal: Clock, value: Boolean): Unit
+  def pokeClock(signal: Clock, value: Boolean): Unit = {
+    val intValue = if (value) 1 else 0
+    simulatorInterface.poke(dataNames(signal), intValue)
+    logger.debug(s"${resolveName(signal)} <- $intValue")
+  }
 
   /** Read the value of a clock.
     */
-  def peekClock(signal: Clock): Boolean
+  def peekClock(signal: Clock): Boolean = {
+    doPeek(signal, new Throwable)
+    val a = simulatorInterface.peek(dataNames(signal)).getOrElse(BigInt(0))
```

Similarly, don't like silent failures.